templates/download_source

```diff
@@ -0,0 +1,141 @@
+#!/bin/bash -eu
+#
+# Copyright (C) 2017 Google Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+readonly BLACKLIST_FILE="/tmp/blacklisted_packages.list"
+readonly SOURCESD_BLACKLIST_FILE="/tmp/blacklisted_sources.list"
+
+function list_all_packages() {
+  dpkg-query -W | cut -f 1 | cut -d ":" -f 1 | sort
+}
+
+function list_packages() {
+  comm -13 <(sort "${BLACKLIST_FILE}") <(list_all_packages)
+}
+
+function list_gpl_packages() {
+  for pkg in $(list_packages); do
+    if egrep -q '\WGPL\W' "/usr/share/doc/${pkg}/copyright"; then
+      echo "${pkg}"
+    fi
+  done
+}
+
+function blacklist_all_current() {
+  list_all_packages >> "${BLACKLIST_FILE}"
+}
+
+function blacklist_nothing() {
+  cat /dev/null > "${BLACKLIST_FILE}"
+}
+
+function blacklist() {
+  for pkg in $@; do
+    echo "${pkg}" >> "${BLACKLIST_FILE}"
+  done
+}
+
+function exclude_sources() {
+  cat /dev/null > "${SOURCESD_BLACKLIST_FILE}"
+  for filename in $@; do
+    echo "${filename}" >> "${SOURCESD_BLACKLIST_FILE}"
+  done
+}
+
+function all_sources_list_files() {
+  for sources_file in /etc/apt/sources.list /etc/apt/sources.list.d/*; do
+    echo "${sources_file}"
+  done
+}
+
+function deb_srcs() {
+  local -r tmp_file="$(mktemp)"
+  for sources_file in \
+    $(comm -13 <(sort "${SOURCESD_BLACKLIST_FILE}") <(all_sources_list_files)); do
+    cat "${sources_file}" | grep -e '^deb ' | sed 's/^deb /deb-src /g' >> "${tmp_file}"
+  done
+  cat "${tmp_file}" | sort | uniq
+}
+
+function download_source_code() {
+  local -r target_dir="$1"
+  local -r src_list_file="/etc/apt/sources.list.d/all-src.list"
+
+  if ! [[ -e "${BLACKLIST_FILE}" ]]; then
+    echo "No package blacklist found. This might download a lot of code."
+    echo "If it is really desired, run with argument 'blacklist-nothing' first."
+    exit 1
+  fi
+
+  # Add current deb entries as deb-src entries from all sources list file,
+  # except the ones in the blacklist.
+  touch "${SOURCESD_BLACKLIST_FILE}"  # Empty blacklist if one doesn't exist.
+  deb_srcs > "${src_list_file}"
+  apt-get update
+
+  cd "${target_dir}"
+  for pkg in $(list_gpl_packages); do
+    echo "--> Downloading source code for GPL package ${pkg}..."
+    apt-get source -q -d "${pkg}"
+  done
+
+  rm "${src_list_file}" "${BLACKLIST_FILE}" "${SOURCESD_BLACKLIST_FILE}"
```

Comment:
We can also use `rm -f` instead of `touch` above.

`touch "${SOURCESD_BLACKLIST_FILE}"  # Empty blacklist if one doesn't exist.`